modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/wal/serializer/RecordDataV1Serializer.java

```diff
@@ -210,22 +208,15 @@ public RecordDataV1Serializer(GridCacheSharedContext cctx) {
     @Override public WALRecord readRecord(RecordType type, ByteBufferBackedDataInput in, int size)
         throws IOException, IgniteCheckedException {
         if (type == ENCRYPTED_RECORD || type == ENCRYPTED_RECORD_V2) {
-            if (encSpi == null) {
-                T2<Integer, RecordType> knownData = skipEncryptedRecord(in, true);
+            EncryptedData encryptedData = readEncryptedData(in, true, type == ENCRYPTED_RECORD_V2);
```

Comment:
It seems a bit weird that a public method of a class related to v1 actually does all the work for both v1 and v2. Probaly, this led to confusion and caused the bug you are fixiing here.

How about the following:

1. Leave the implementation of `readRecord()` here, but make it a template method
2. Extract things which define the differences between v1 and v2 to protected methods and overriding them in `RecordDataV2Serializer`?

For example, the template method could look like this:

```
    @Override public WALRecord readRecord(RecordType type, ByteBufferBackedDataInput in, int size)
        throws IOException, IgniteCheckedException {
        if (type == encryptedRecordType()) {
            EncryptedData encryptedData = readEncryptedData(in, true);

            if (encryptedData.isDecrypted()) {
                ByteBufferBackedDataInput data = encryptedData.data();

                return readPlainRecord(encryptedData.recordType(), data, true, data.buffer().capacity());
            }
            else
                return new EncryptedRecord(encryptedData.grpId(), encryptedData.recordType());
        }

        return readPlainRecord(type, in, false, size);
    }
```

and there would also be

```
protected RecordType encryptedRecordType();
```

returning `ENCRYPTED_RECORD` and `ENCRYPTED_RECORD_V2` in the corresponding implementations, and

```
    protected EncryptedData readEncryptedData(
        ByteBufferBackedDataInput in,
        boolean readType
    ) throws IOException, IgniteCheckedException {
```

calling `readEncryptedData(ByteBufferBackedDataInput, boolean, boolean)` with the corresponding third argument specific for v1 or v2.

WDYT?

Reply:
Nevermind, the hierarchy of record serializer versions is orthogonal to the hierarchy of message versions

Reply:
I don't think I fully understand your suggestion: `ENCRYPTED_RECORD` and `ENCRYPTED_RECORD_V2` are not serialized solely by corresponding serializers, but rather by both of them.